p2panda-rs/src/secret_group/group.rs

```diff
@@ -0,0 +1,552 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use openmls::framing::{MlsMessageIn, MlsMessageOut, VerifiableMlsPlaintext};
+use openmls::group::GroupId;
+use openmls::prelude::KeyPackage;
+use openmls_traits::OpenMlsCryptoProvider;
+use tls_codec::{Deserialize, Serialize, TlsVecU32};
+
+use crate::hash::Hash;
+use crate::secret_group::lts::{
+    LongTermSecret, LongTermSecretCiphersuite, LongTermSecretEpoch, LongTermSecretNonce,
+    LTS_DEFAULT_CIPHERSUITE, LTS_EXPORTER_LABEL, LTS_NONCE_EXPORTER_LABEL,
+};
+use crate::secret_group::mls::MlsGroup;
+use crate::secret_group::{
+    SecretGroupCommit, SecretGroupError, SecretGroupMember, SecretGroupMessage,
+};
+
+type LongTermSecretVec = TlsVecU32<LongTermSecret>;
+
+/// Create or join secret groups, maintain their state and en- / decrypt user messages.
+#[derive(Debug)]
+pub struct SecretGroup {
+    /// Used ciphersuite when generating new long term secrets
+    long_term_ciphersuite: LongTermSecretCiphersuite,
+
+    /// Internal counter for AEAD nonce.
+    long_term_nonce: LongTermSecretNonce,
+
+    /// Stored long term secrets (AEAD keys).
+    long_term_secrets: LongTermSecretVec,
+
+    /// Messaging Layer Security (MLS) group.
+    mls_group: MlsGroup,
+
+    /// Flag indicating if group was created by us.
+    owned: bool,
+}
+
+impl SecretGroup {
+    // Creation
+    // ========
+
+    /// Creates a new `SecretGroup` instance which can be used to encrypt data securely between
+    /// members of the group.
+    ///
+    /// The first member of this group will automatically be the "owner" maintaining the group
+    /// state by updating secrets, adding or removing members.
+    pub fn new(
+        provider: &impl OpenMlsCryptoProvider,
+        group_instance_id: &Hash,
+        member: &SecretGroupMember,
+    ) -> Result<Self, SecretGroupError> {
+        // Generate new InitKeys and consume them directly when creating MLS group
+        let init_key_package = member.key_package(provider)?;
+
+        // Internally we use the MLS `GroupId` struct to represent groups since it already
+        // implements the TLS encoding traits
+        let group_id = GroupId::from_slice(&group_instance_id.to_bytes());
+
+        // Create the MLS group with first member inside
+        let mls_group = MlsGroup::new(provider, group_id, init_key_package)?;
+
+        let mut group = Self {
+            // Hard code long term secret ciphersuite for now
+            long_term_ciphersuite: LTS_DEFAULT_CIPHERSUITE,
+            long_term_nonce: LongTermSecretNonce::default(),
+            long_term_secrets: Vec::new().into(),
+            mls_group,
+            owned: true,
+        };
+
+        // Generate first long term secret and store it in secret group
+        group.rotate_long_term_secret(provider)?;
+
+        Ok(group)
+    }
+
+    /// Creates a `SecretGroup` instance by joining an already existing group.
+    pub fn new_from_welcome(
```

Comment:
> Is the group_instance_id contained in the welcome message here?

Yes, it is part of it! See: https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#section-11.2.2 (oh, there is new version).

> I am not sure whether I understand how to use this: How do I know which commit to feed into this to get the group instance? Is it always the same commit (the welcome with which I was added to the group) or always the latest group commit?

In a "classical" MLS scenario the Delivery Service would make sure the Welcome message arrives to the members computer directly. In our case it is a little bit different. The flow could then look like this:

**Client**

1. I'm starting a p2panda client
2. I'm downloading instances via GraphQL of something this client is interested in
3. I detect that some values of these instances are encrypted
4. I read what `SecretGroup` instance these ciphertexts belong to
5. I query which commit addresses me, download all `SecretGroupCommits` from that point on
6. I process the Welcome message via `new_from_welcome` in that `SecretGroupCommit` which addresses me
7. I decrypt the instance field value

**Node**

The Welcome message contains information about who gets added (see `KeyPackageRef` in Section 11.2.2).

1. During materialization the node would look at every `SecretGroupCommit`
2. Check if its an Add, Remove or Update Commit
3. If its an Add Commit, it would look into the Welcome message and attached `KeyPackageRef` values
4. It would store these information in the database and make them queryable

Reply:
> The Welcome message contains information about who gets added (see KeyPackageRef in Section 11.2.2).

This answers my question! Thank you, also for the full overview, quite helpful.